### PR TITLE
[feature] search chats

### PR DESCRIPTION
**Pull request by yenda**
search4

screenshot_1550177662
screenshot_1550177669
screenshot_1550177674
screenshot_1550214493
screenshot_1550214528

status: ready 

# Note for testing

check if desktop is not broken

### CODE REVIEW

**Comment by flexsurfer** [comment]
Nice!

**Comment by flexsurfer** [comment]
Is it possible to hide it when scroll down and search is empty?

**Comment by yenda** [comment]
@flexsurfer it hides when you press cancel

File: src/status_im/ui/screens/home/views.cljs [hunk 1]
```diff
@@ -87,15 +91,160 @@
...
+                                                     (re-frame/dispatch [:search/filter-changed nil])
+                                                     (hide-search-input anim-search-input-height))
+                                       :on-focus #(when-not search-filter
+                                                    (re-frame/dispatch [:search/filter-changed ""]))
+                                       :on-change (fn [text]
+                                                    (re-frame/dispatch [:search/filter-changed text]))}]]
+         (case search-filter
+           nil [list/flat-list {:data      all-home-items
+                                :key-fn    first
+                                :onScrollBeginDrag (fn [e]
```
Comment (flexsurfer) [inline src/status_im/ui/screens/home/views.cljs]: kebab?

File: src/status_im/ui/screens/home/views.cljs [hunk 2]
```diff
@@ -87,15 +91,160 @@
...
+(def search-input-height 56)
+
+(defn search-input
+  [search-filter {:keys [on-cancel on-focus on-change]}]
+  (let [input-is-focused? (reagent/atom false)
+        input-ref (reagent/atom nil)]
+    (fn [search-filter]
+      (let [show-cancel? (or @input-is-focused?
+                             (not-empty search-filter))]
+        [react/view {:style {:height search-input-height
```
Comment (flexsurfer) [inline src/status_im/ui/screens/home/views.cljs]: maybe move these to styles

File: src/status_im/ui/screens/home/views.cljs [hunk 3]
```diff
@@ -2,7 +2,10 @@
   (:require-macros [status-im.utils.views :as views])
```
Comment (flexsurfer) [inline src/status_im/ui/screens/home/views.cljs]: maps values formatting

**Review by flexsurfer — APPROVED** [review]

File: src/status_im/ui/screens/home/views.cljs [hunk 4]
```diff
@@ -87,15 +91,165 @@
...
+(def search-input-height 56)
+
+(defn search-input
+  [search-filter {:keys [on-cancel on-focus on-change]}]
+  (let [input-is-focused? (reagent/atom false)
+        input-ref (reagent/atom nil)]
+    (fn [search-filter]
+      (let [show-cancel? (or @input-is-focused?
+                             (not-empty search-filter))]
+        [react/view {:style {:height             search-input-height
```
Comment (vitvly) [inline src/status_im/ui/screens/home/views.cljs]: Probably would make sense to separate away styles so that to aid readability.

File: src/status_im/ui/screens/home/views.cljs [hunk 5]
```diff
@@ -87,15 +91,165 @@
...
+                                [react/text {:style {:font-size   15
+                                                     :margin-left 16
+                                                     :margin-top  16
+                                                     :color       colors/gray}}
+                                 (i18n/label title)])
+    :render-section-footer-fn (fn [{:keys [title data]}]
+                                (when (empty? data)
+                                  (case title
+                                    "messages"
+                                    [list/big-list-item {:text          (i18n/label :t/messages-search-coming-soon)
```
Comment (vitvly) [inline src/status_im/ui/screens/home/views.cljs]: Seems that all these entries are identical except for the icon keyword and text label in the first element, so would be better to move `case` expression so that it applies to `[:text :icon]` keys only.

**Review by vitvly — APPROVED** [review]
Sweeeeet.

File: src/status_im/ui/screens/home/views.cljs (hunk 4 above)
Comment (yenda) [reply src/status_im/ui/screens/home/views.cljs]: the style isn't reusable and since it is a view it is mostly concerned with how things should be rendered, I find it much easier to adjust styling when things are in one place, also when it comes to hunting down useless layers of components

Comment (vitvly) [reply src/status_im/ui/screens/home/views.cljs]: Got it, agree.

**Comment by churik** [comment]
Link to last IOS build: https://ci.status.im/job/status-react/job/prs/job/ios/job/PR-7511/lastSuccessfulBuild/artifact/StatusIm-190215-115657-eba34c-pr.ipa

@yenda 
### 1) 'Search' is shown when chat list is empty (IOS, Android)
**Steps:**
- create new account
- go to wallet and back

**Expected result:** no 'Search' is shown
**Actual result:** 
s1 


### 2) Can't set focus in 'Search' on IOS (search is unacessible) 

**Steps:**
- create new account
- join several chats
- try to search smth (pull down to get search visible)

**Expected result:** can set focus in 'Search'
**Actual result:**  'Search' is hiding and page is scrolling back
**Video:**  https://take.ms/LNAdC


### 3) Not possible to open search when you have several chats (less amount than is required for scroll bar) (Android)

**Steps:**
- create new account
- join 2 chats
- try to search smth (pull down to get search visible)

**Expected result:** can get 'Search' visibe
**Actual result:**  'Search' is not visible after pulling down
So for Android in this case 'Search' is not shown
On IOS in this case part of chat is overlapping, video: https://take.ms/yTw7j

### 4) User input is not preserved when user taps outside "Search" field set focus again in "Search"

**Steps:**
- join 10 chats
- try to search smth (pull down to get search visible)
- tap outside "Search"
- tap in "Search" field

**Expected result:** user input is preserved
**Actual result:**  user input is erased
**Video:** https://take.ms/xZChg


### 5) Last chat in chat list is overlapped with bottom panel
**Steps:**
- join 10 chats
- open search field (pull down to get search visible)
- scroll down to last chat in lst

**Expected result:** can see last chat
**Actual result:** 
s4


### 6) Bottom panel appears for a moment and disappears when set focus on search
**Steps:**
- join 10 chats
- open search field (pull down to get search visible)
- tap on search field to set focus

**Expected result:** bottom panel doesn't appear
**Actual result:** 
s6
**Video:** https://take.ms/Juyor (on 4 sec)

### 7) Home becomes unresponsive after search for non-existing chat, switching to another tab and coming back

**Steps:**
- join 10 chats
- open search field (pull down to get search visible)
- search for smth non-existing, i.e. `____`
- tap on wallet
- come back to nome screen

**Expected result:** can cancel search
**Actual result:** screen is unresponsive, can't tap on "Search" or on cancel
**Video:** https://take.ms/EGIHV

**Comment by oskarth** [comment]
Thanks for adding the video and screenshots @yenda! Useful for sanity check during review.

@churik in your opinion, is there something in your testing here that you think should've been caught before QA had a look at it? Perhaps there's a set of tools /question that would be useful for PRs to be sanity checked against before asking QA to test. This might involve fewer PRs with 'tested-issues'.

cc @annadanchenko @PombeirP @flexsurfer

**Comment by churik** [comment]
@oskarth thanks for the question - I believe items 1,2,3 can be caught before QA look for sure (it is just basic usage IMO) - so it took ~ 5-10 mins to find these issues and around an hour to get all screenshots and make descriptions.

**Comment by yenda** [comment]
@churik everything should be fixed. Just not sure about 6, couldn't reproduce but it is more related to the bottom bar than this PR

**Comment by churik** [comment]
@yenda 
**1) When chat list is empty, empty screen is shown instead of  "...Tap the plus button to get started"**
**Steps:**
- create new account
- go to wallet and back

**Expected result:** no 'Search' is possible to make on empty chats; "...Tap the plus button to get started" message is shown
**Actual result:** empty screen, can open "Search" when pull down
empl

**2, 3, 4, 5, 7) fixed**

**6) It 's not a pr problem I also saw this in another PR. So agree here.**

**8) When you have 1 chat in chat list empty list is shown after you set input to "Search"**
It is not issue, mostly like feature request. Don't understand why empty chat list is shown when you set focus to search field, but don't start typing. Checked in several other chats with similar functionality - everywhere chat list is shown in this case.
Steps: 
- create new account
- join to several chats
- pull down 
- set focus to "Search" field

Expected result: can see chat list until start typing
Actual result: empty chat list
empt chu


**9) Errors below chat header are overlapping when tap on "Cancel"**

Steps: 
- create new account
- join to one chat
- pull down to have "Search"opened
- go to offline
- tap cancel near "Search"

Expected result: "Offline" message is shown, no search field is shown
Actual result: 
obserb

**10)  Can't scroll when keyboard is opened and focus is set on "Search" field (IOS only)**
Prerequisites: have several chats
Steps: 
- search for smth that will show any results. Don't close keyboard.
- try to scroll 

Expected result: can scroll
Actual result: can't scroll to the end of list
ezgif com-video-to-gif


**11.1) "Chat", "Browser", "Messages" labels are ovelapping with list of results when scrolling**

Prerequisites: have several chats
Steps:
- search for smth that will show any results
- scroll results
Expected result: no overlapping
Actual result: 
hdhdh

**11.2) "Search message history is coming soon" should be displayed fully (minor)**
What it looks now: 
not_full
And IMO this message doesn't provide any useful info.

**12) "Error is occured after fetching history" is overlapping with search button (minor)**
Basically if error message contains > 1 row, it is overlapping partially panel. Probably it is not related to current PR, because it lloks cut from the botton on nightly as well.
eerr

**Comment by yenda** [comment]
@churik @errorists unless I'm mistaken 8 is the normal behavior from the design (and was a bit painful to implement, but really not as much as hiding the search)

**Comment by errorists** [comment]
> @churik @errorists unless I'm mistaken 8 is the normal behavior from the design (and was a bit painful to implement, but really not as much as hiding the search)

Yes, 8 is by design, we could either not show the chat list at all or animate it to a reduced opacity so the focus is clearly on search, of those two I picked the simpler approach.

**Comment by yenda** [comment]
@churik I've fixed and test the different repro steps on Android OnePlus5T and iPhone 6S for 1, 9, 10 and 11. According to @errorists the other 11 (ellipses at the end of label) are ok for now but we can improve later by having 2 lines. Can you create an issue for 12?

**Comment by churik** [comment]
@yenda 
retested all items, looks unfixed for me only 10.

**10) Can't scroll when keyboard is opened and focus is set on "Search" field (IOS only)**
Prerequisites: have several chats + DApps
Steps:

* search for smth that will show any results. Don't close keyboard.
* try to scroll

Expected result: can scroll
Actual result: can't scroll to the end of list (result for Browser can't be seen unless you close a keyboard)
I'm on IOS 11.4.1.
If it is not reproducible for you I can ask other QA to check it on other devices.

**Comment by churik** [comment]
Tested on IOS 11.4.1 (IPhone 7) and LG v20 (Android 8):
- on empty chat list;
- with invalid input;
- with results of all types: Dapps, 1-1, public, group
- autologin with filled search
- no overlapping with system messages

Waiting for e2e, lgtm


@antdanchenko test case is added for e2e: https://ethstatus.testrail.net/index.php?/cases/view/5757
